kmip.h

```diff
@@ -1092,6 +1183,84 @@ typedef struct response_message
+} ApplicationNamespaces;
+*/
+
+typedef struct query_request_payload
+{
+    LinkedList *functions;
+} QueryRequestPayload;
+
+typedef struct query_response_payload
+{
+    Operations*             operations;              // Specifies an Operation that is supported by the server.
+    ObjectTypes*            objects;                 // Specifies a Managed Object Type that is supported by the server.
+    TextString*             vendor_identification;   // SHALL be returned if Query Server Information is requested. The Vendor Identification SHALL be a text string that uniquely identifies the vendor.
+    ServerInformation*      server_information;      // Contains vendor-specific information possibly be of interest to the client.
+ // ApplicationNamespaces*  application_namespaces;  // Specifies an Application Namespace supported by the server.
+ // Extension Information         No, MAY be repeated  // SHALL be returned if Query Extension List or Query Extension Map is requested and supported by the server.
+ // Attestation Type              No, MAY be repeated  // Specifies an Attestation Type that is supported by the server.
+ // RNG Parameters                No, MAY be repeated  // Specifies the RNG that is supported by the server.
+ // Profile Information           No, MAY be repeated  // Specifies the Profiles that are supported by the server.
+ // Validation Information        No, MAY be repeated  // Specifies the validations that are supported by the server.
+ // Capability Information        No, MAY be repeated  // Specifies the capabilities that are supported by the server.
+ // Client Registration Method    No, MAY be repeated  // Specifies a Client Registration Method that is supported by the server.
+ // Defaults Information          No                   // Specifies the defaults that the server will use if the client omits them.
+ // Protection Storage Masks      Yes                  // Specifies the list of Protection Storage Mask values supported by the server. A server MAY elect to provide an empty list in the Response if it is unable or unwilling to provide this information.
+} QueryResponsePayload;
+
+
+#define MAX_QUERY_LEN    128
+#define MAX_QUERY_OPS   0x40
+#define MAX_QUERY_OBJS  0x20
+
+typedef struct query_response
+{
+    size_t           operations_size;
+    int              operations[MAX_QUERY_OPS];
+    size_t           objects_size;
+    int              objects[MAX_QUERY_OBJS];
+    char             vendor_identification[MAX_QUERY_LEN];
+    bool32           server_information_valid;
+    char             server_name[MAX_QUERY_LEN];
+    char             server_serial_number[MAX_QUERY_LEN];
+    char             server_version[MAX_QUERY_LEN];
+    char             server_load[MAX_QUERY_LEN];
+    char             product_name[MAX_QUERY_LEN];
+    char             build_level[MAX_QUERY_LEN];
+    char             build_date[MAX_QUERY_LEN];
+    char             cluster_info[MAX_QUERY_LEN];
+} QueryResponse;
```

Comment:
The fields in this struct don' correspond to the fields defined in the specification. Defining maximum array lengths also doesn't align with the repeating nature of some of these fields. Most of them should be linked lists that grow as needed to fit the encoding or user usage.